opentelemetry_api/opentelemetry/context/traceparent.c

```diff
@@ -0,0 +1,106 @@
+/* Copyright 2019, OpenTelemetry Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "traceparent.h"
+
+status_t _uint_from_hex_string(unsigned int* value, const char* s) {
+  char c = s[0];
+  unsigned int result;
+
+  if (c >= '0' && c <= '9')
+  {
+    result = c - '0';
+  }
+  else if (c >= 'a' && c <= 'f')
+  {
+    result = c - 'a' + 10;
+  }
+  else
+  {
+    return STATUS_ERROR_INPUT_ILLEGAL;
+  }
+  result <<= 4;
+  c = s[1];
+  if (c >= '0' && c <= '9')
+  {
+    result += c - '0';
+  }
+  else if (c >= 'a' && c <= 'f')
+  {
+    result += c - 'a' + 10;
+  }
+  else
+  {
+    return STATUS_ERROR_INPUT_ILLEGAL;
+  }
+  *value = result;
+
+  return STATUS_OK;
+}
+
+status_t _trace_parent_from_string(TraceParent* trace_parent, const char* s)
+{
+  unsigned int value;
+  /* if this function failed, trace_parent could be polluted,
+     do we want to take a copy (perf sucks)?
+   */
+  IF_ERROR_RETURN(_uint_from_hex_string(&value, s));
+  trace_parent->version = value;
+  s += 2;
+  IF_FALSE_RETURN('-' == *s, STATUS_ERROR_INPUT_ILLEGAL)
+  s += 1;
+  for (int i = 0; i < 16; i++)
+  {
+    IF_ERROR_RETURN(_uint_from_hex_string(&value, s));
+    trace_parent->trace_id[i] = value;
+    s += 2;
+  }
+  IF_FALSE_RETURN('-' == *s, STATUS_ERROR_INPUT_ILLEGAL)
+  s += 1;
+  for (int i = 0; i < 8; i++)
+  {
+    IF_ERROR_RETURN(_uint_from_hex_string(&value, s));
+    trace_parent->parent_id[i] = value;
+    s += 2;
+  }
+  IF_FALSE_RETURN('-' == *s, STATUS_ERROR_INPUT_ILLEGAL)
+  s += 1;
+  IF_ERROR_RETURN(_uint_from_hex_string(&value, s));
+  trace_parent->trace_flags = value;
+  s += 2;
+  IF_FALSE_RETURN(('-' == *s || ('\0' == *s)), STATUS_ERROR_INPUT_ILLEGAL)
+  return STATUS_OK;
+}
+
+status_t _trace_parent_to_string(const TraceParent* trace_parent, char* s)
+{
+  return STATUS_ERROR_UNKNOWN;
+}
+
+status_t trace_parent_from_string(TraceParent* trace_parent, const char* s)
```

Comment:
Personally I'm in favor of `PascalCase` instead of `snake_case`. Should we vote? 😄

Reply:
I think we should base these functions off of a back-ported string_view instead of using null terminated strings (i.e. we copy an implementation of string_view that's compatible with c++17's string_view; we can even opt in to std::string_view like [absl](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/string_view.h#L33) if they're using c++17+).

Lot's of systems don't use null terminated strings (for example [nginx](http://nginx.org/en/docs/dev/development_guide.html#strings), [lua](https://www.lua.org/pil/24.2.1.html))

So having the api require a null-terminated c-style string would require them to make a copy before calling the function.

Reply:
> Personally I'm in favor of PascalCase instead of snake_case

Instead of debating convention on a case-by-case basis, could we use #5 to adopt one of the standard naming conventions globally and then redo the code to match that.

Reply:
I'm very much in favor of string_view.

Also in favor of PascalCase, we can discuss that in #5 if it's more appropriate.